PopPUNK/lineage_clustering.py

```diff
@@ -0,0 +1,680 @@
+        
+    # iterate through the ref-query distances
+    for index,(distance,(isolate1,isolate2)) in enumerate(zip(query_distances,query_row_names)):
+        # identify ref-query matches
+        ref = None
+        query = None
+        if isolate1 in max_distance.keys() and isolate2 not in max_distance.keys():
+            ref = isolate1
+            query = isolate2
+        elif isolate2 in max_distance.keys() and isolate1 not in max_distance.keys():
+            ref = isolate2
+            query = isolate1
+        if ref is not None:
+            if distance <= max_distance[ref]:
+                # unset isolate and references
+                lineage_clustering[ref] = null_cluster_value
+                for neighbour in nn[ref]:
+                    lineage_clustering[neighbour] = null_cluster_value
+                # add neighbours
+                nn[ref][query] = distance
+                # delete links if no longer high ranked match
+                if distance < max_distance[ref]:
+                    if num_distances[ref] == R:
+                        to_delete = []
+                        for other in nn[ref].keys():
+                            if nn[ref][other] == max_distance[ref]:
+                                to_delete.append(other)
+                        for other in to_delete:
+                            del nn[ref][other]
+                    else:
+                        # if set from query-query distances < R
+                        num_distances[ref] = num_distances[ref] + 1
+                    max_distance[ref] = max(nn[ref].values())
+    # return updated dict
+    return nn, lineage_clustering
+
+def cluster_into_lineages(X, rank_list = None, output = None, rlist = None, qlist = None, existing_scheme = None, use_accessory = False, num_processes = 1):
+    """ Clusters isolates into lineages based on their
+    relative distances.
+    
+    Args:
+        X (np.array)
+            n x 2 array of core and accessory distances for n samples.
+            This should not be subsampled.
+        rank_list (list of int)
+            Integers specifying the maximum rank of neighbours used
+            for clustering.
+        rlist (list)
+            List of reference sequences.
+        qlist (list)
+            List of query sequences being added to an existing clustering.
+            Should be included within rlist.
+        existing_scheme (str)
+            Path to pickle file containing lineage scheme to which isolates
+            should be added.
+        use_accessory (bool)
+            Option to use accessory distances rather than core distances.
+        num_processes (int)
+            Number of CPUs to use for calculations.
+            
+    Returns:
+        combined (dict)
+            Assignment of each isolate to clusters by all ranks used.
+    """
+    
+    # process distance matrix
+    # - this should be sorted (PyTorch allows this to be done on GPUs)
```

Comment:
It wouldn't be too hard to return a (GPU) sorted query from sketchlib

Reply:
np.argsort is good though, could be parallelise by splitting matrix into blocks of rows